executor/internal/mpp/local_mpp_coordinator.go

```diff
@@ -0,0 +1,486 @@
+			zap.Uint64("ServerID", mppTask.MppQueryID.ServerID), zap.String("address", mppTask.Meta.GetAddress()),
+			zap.String("plan", plannercore.ToString(pf.ExchangeSender)),
+			zap.Int64("mpp-version", mppTask.MppVersion.ToInt64()),
+			zap.String("exchange-compression-mode", pf.ExchangeSender.CompressionMode.Name()),
+		)
+		req := &kv.MPPDispatchRequest{
+			Data:       pbData,
+			Meta:       mppTask.Meta,
+			ID:         mppTask.ID,
+			IsRoot:     pf.IsRoot,
+			Timeout:    10,
+			SchemaVar:  c.is.SchemaMetaVersion(),
+			StartTs:    c.startTS,
+			MppQueryID: mppTask.MppQueryID,
+			MppVersion: mppTask.MppVersion,
+			State:      kv.MppTaskReady,
+		}
+		c.mppReqs = append(c.mppReqs, req)
+	}
+	return nil
+}
+
+func (c *localMppCoordinator) dispatchAll(ctx context.Context) {
+	for _, task := range c.mppReqs {
+		if atomic.LoadUint32(&c.closed) == 1 {
+			break
+		}
+		c.mu.Lock()
+		if task.State == kv.MppTaskReady {
+			task.State = kv.MppTaskRunning
+		}
+		c.mu.Unlock()
+		c.wg.Add(1)
+		boMaxSleep := copr.CopNextMaxBackoff
+		failpoint.Inject("ReduceCopNextMaxBackoff", func(value failpoint.Value) {
+			if value.(bool) {
+				boMaxSleep = 2
+			}
+		})
+		bo := backoff.NewBackoffer(ctx, boMaxSleep)
+		go func(mppTask *kv.MPPDispatchRequest) {
+			defer func() {
+				c.wg.Done()
+			}()
+			c.handleDispatchReq(ctx, bo, mppTask)
+		}(task)
+	}
+	c.wg.Wait()
+	close(c.wgDoneChan)
+	close(c.respChan)
+}
+
+func (c *localMppCoordinator) sendError(err error) {
+	c.sendToRespCh(&mppResponse{err: err})
+	c.cancelMppTasks()
+}
+
+func (c *localMppCoordinator) sendToRespCh(resp *mppResponse) (exit bool) {
+	defer func() {
+		if r := recover(); r != nil {
+			logutil.BgLogger().Error("localMppCoordinator panic", zap.Stack("stack"), zap.Any("recover", r))
+			c.sendError(errors.New(fmt.Sprint(r)))
+		}
+	}()
+	if c.memTracker != nil {
+		respSize := resp.MemSize()
+		failpoint.Inject("testMPPOOMPanic", func(val failpoint.Value) {
+			if val.(bool) && respSize != 0 {
+				respSize = 1 << 30
+			}
+		})
+		c.memTracker.Consume(respSize)
+		defer c.memTracker.Consume(-respSize)
+	}
+	select {
+	case c.respChan <- resp:
+	case <-c.finishCh:
+		exit = true
+	}
+	return
+}
+
+// TODO:: Consider that which way is better:
+// - dispatch all Tasks at once, and connect Tasks at second.
+// - dispatch Tasks and establish connection at the same time.
+func (c *localMppCoordinator) handleDispatchReq(ctx context.Context, bo *backoff.Backoffer, req *kv.MPPDispatchRequest) {
+	var rpcResp *mpp.DispatchTaskResponse
+	var err error
+	var retry bool
+	for {
+		rpcResp, retry, err = c.sessionCtx.GetMPPClient().DispatchMPPTask(
+			kv.DispatchMPPTaskParam{
+				Ctx:                        ctx,
+				Req:                        req,
+				EnableCollectExecutionInfo: c.enableCollectExecutionInfo,
+				Bo:                         bo.TiKVBackoffer(),
+			})
+		if retry {
```

Comment:
Maybe add a todo to improve the retry mechanism of mpp, such as https://github.com/pingcap/tidb/issues/31015

Reply:
Not sure if I get your point, add a todo comment here. Please double check it.

Reply:
yes, that is my point